fdbserver/BlobManifest.actor.cpp

```diff
@@ -90,24 +108,220 @@ struct BlobManifestFile {
+			}
+
+			// return the manifest if it's valid
+			BlobManifest manifest(result);
+			if (manifest.isValid()) {
+				return manifest;
+			} else {
+				dprint("Skip corrupted manifest {} {}\n", firstFile.epoch, firstFile.seqNo);
+				result.clear(); // prepare for next search
+			}
+		}
+
+		dprint("No valid blob manifest files\n");
+		TraceEvent("BlobRestoreMissingManifest").detail("Url", SERVER_KNOBS->BLOB_RESTORE_MANIFEST_URL);
+		throw blob_restore_missing_manifest();
+	}
+
+	// Delete all files of oldest manifest
+	static void deleteOldest(const std::vector<BlobManifestFile>& allFiles,
+	                         Reference<BackupContainerFileSystem> container) {
+		if (allFiles.empty()) {
+			return;
+		}
+		int64_t epoch = allFiles.back().epoch;
+		int64_t seqNo = allFiles.back().seqNo;
+		for (auto& f : allFiles) {
+			if (f.epoch == epoch && f.seqNo == seqNo) {
+				container->deleteFile(f.fileName);
+			}
+		}
+	}
+
+	// Count how many manifests
+	static int count(const std::vector<BlobManifestFile>& allFiles) {
+		if (allFiles.empty())
+			return 0;
+
+		int64_t epoch = allFiles.front().epoch;
+		int64_t seqNo = allFiles.front().seqNo;
+		int count = 1;
+		for (auto& f : allFiles) {
+			if (f.epoch != epoch || f.seqNo != seqNo) {
+				count++;
+				epoch = f.epoch;
+				seqNo = f.seqNo;
+			}
+		}
+		return count;
+	}
+
+private:
+	std::vector<BlobManifestFile> files_;
+};
+
+// Splitter could write manifest content into a collection of files with bounded # rows
+class BlobManifestFileSplitter : public ReferenceCounted<BlobManifestFileSplitter> {
+public:
+	BlobManifestFileSplitter(Reference<BlobConnectionProvider> blobConn, int64_t epoch, int64_t seqNo)
+	  : segmentNo_(1), blobConn_(blobConn), epoch_(epoch), seqNo_(seqNo), closed_(false), totalRows_(0) {}
+
+	// Append a new row to the splitter
+	ACTOR static Future<Void> append(Reference<BlobManifestFileSplitter> self, KeyValueRef row) {
+		ASSERT(!self->closed_);
+		self->rows_.push_back_deep(self->rows_.arena(), row);
+		++self->totalRows_;
+		if (self->rows_.size() > SERVER_KNOBS->BLOB_RESTORE_MANIFEST_FILE_MAX_ROWS) {
+			wait(self->flushNext(self));
+		}
+		return Void();
+	}
+
+	// Close the splitter. No more data should be added after
+	ACTOR static Future<Void> close(Reference<BlobManifestFileSplitter> self) {
+		wait(flushNext(self));
+		wait(flushTailer(self));
+		self->closed_ = true;
+		return Void();
+	}
+
+	// Reset writer and start to write from the beginning
+	void reset() {
+		dprint("Reset manifest file {} {}\n", epoch_, seqNo_);
+		TraceEvent("ResetBlobManifestFile").detail("SegNo", segmentNo_);
+		rows_.clear();
+		segmentNo_ = 1;
+		totalRows_ = 0;
+	}
+
+private:
+	// Write next segment
+	ACTOR static Future<Void> flushNext(Reference<BlobManifestFileSplitter> self) {
+		Standalone<BlobManifestSegment> segment;
+		segment.rows = self->rows_;
+		segment.segmentNo = self->segmentNo_;
+
+		BinaryWriter wr(IncludeVersion(ProtocolVersion::withBlobGranuleFile()));
```

Comment:
We may want to consider reusing the blob granule snapshot file format, as it already takes a set of key-value pairs and serializes them into chunks with encryption and compression. The chunking in particular is ideal for a cooperative multitasking system like fdb where an expensive actor must explicitly yield to other actors.
The only issue I see with directly reusing the snapshot file format that it has the requirement that the key-value pairs are sorted.
Maybe we could either relax that requirement in this case or just reuse parts of the code.

Reply:
as discussed offline, I would reuse snapshot serialization/deserialization for manifest files. already updated. thank you!